elastic/beats
# Osquerybeat: Bundle osquery.app for macOS instead of osquery-*.pkg. Eliminates runtime dependency on pkgutil.

## What does this PR do?

* Changes how osquerybeat macOS bundles osquery.app. It extracts osquery.app bundle and certs files from the official .pgk during the build and includes the into osquerybeat distro package. Osquerybeat was shipping with .pkg initially since there was no good Go library that supported extracting files from .pkg file.
* Adds limited implementation of xar and cpio readers and pkgutil in order to be able to extract the files during build process.
* Adds "PreserveMode" flag for the PackageFile spec in order to be able to add the osquery.app bundle directory and preserve the original permissions on the files in the bundle. This fixes the issue where the specified "Mode" was applied to all the files in the app bundle directory while the app bundle contains a mix of files with and without executable permissions.

## Why is it important?

* Removes runtime dependency on pkgutil.
* Eliminates the "install" step for osquery.app on the first run of osquerybeat.
* Bundle only files that we need from osquery distro, same as on for windows and linux osquerybeat distribution.

## Checklist


## How to test this PR locally

Run the agent with osquery manager, make sure it works and osquery returns livequery results.

## Screenshots

The agent components directory before the change, version 8.6.0 for example:
<img width="649" alt="Screenshot 2023-02-13 at 3 44 04 PM" src="https://user-images.githubusercontent.com/872351/218575563-09d1032c-7987-4014-947c-c94a9faf025d.png">

The agent components directory after this change, the osquery.app is in the bundle instead of osquery-*.pkg file:
<img width="647" alt="Screenshot 2023-02-13 at 3 45 11 PM" src="https://user-images.githubusercontent.com/872351/218575883-cb3a48ac-0632-45e8-b90d-e126e100f4a2.png">

Live queries still work:
<img width="1296" alt="Screenshot 2023-02-13 at 3 33 05 PM" src="https://user-images.githubusercontent.com/872351/218575976-2ba44809-0506-4a3d-ad0e-03681f4344f2.png">
